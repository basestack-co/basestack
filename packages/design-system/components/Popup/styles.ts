import { rem } from "polished";
import styled from "styled-components";
import { position } from "styled-system";

export const Container = styled.div<{ width?: number }>`
  ${position};
  background-color: ${({ theme }) => theme.popup.backgroundColor};
  box-shadow: ${({ theme }) => theme.shadow.elevation6};
  padding: ${({ theme }) => theme.spacing.s1};
  border-radius: 4px;
  width: ${({ width }) => (width ? rem(`${width}px`) : rem("150px"))};
  z-index: ${({ theme }) => theme.zIndex.popup};
`;

export const List = styled.ul`
  display: flex;
  flex-direction: column;
`;

export const ListItem = styled.li`
  display: flex;
  flex-direction: column;
`;
