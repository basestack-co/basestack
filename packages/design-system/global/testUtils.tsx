import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
