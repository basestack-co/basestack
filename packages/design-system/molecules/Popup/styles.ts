import styled from "styled-components";
import { rem } from "polished";
import { position } from "styled-system";

export const Container = styled.div`
  ${position};
  background-color: ${({ theme }) => theme.colors.white};
  box-shadow: ${({ theme }) => theme.shadow.elevation6};
  padding: ${({ theme }) => theme.spacing.s1};
  border-radius: 4px;
  width: ${rem("150px")};
  z-index: ${({ theme }) => theme.zIndex.popup};
`;

export const List = styled.ul`
  display: flex;
  flex-direction: column;
`;

export const ListItem = styled.li`
  display: flex;
  flex-direction: column;
`;
