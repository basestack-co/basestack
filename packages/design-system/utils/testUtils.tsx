import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
