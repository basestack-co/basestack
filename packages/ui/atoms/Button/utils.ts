import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
