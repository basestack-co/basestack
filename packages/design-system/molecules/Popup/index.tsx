import { forwardRef, memo } from "react";
import { PositionProps } from "styled-system";
import { Button, ButtonProps, ButtonVariant } from "../../atoms";
import { Container, List, ListItem } from "./styles";

export interface PopupItems {
  /**
   * onClick list item callback
   */
  onClick: () => void;
  /**
   * List item text
   */
  text: string;
  /**
   * Optional List item icon
   */
  icon?: string;
  /**
   * Change list item button styles based on variant
   */
  variant?: ButtonVariant;
}

export interface PopupProps extends PositionProps {
  /**
   * List of actions
   */
  items: Array<PopupItems>;
}

const Popup = forwardRef<HTMLDivElement, PopupProps>(
  ({ items, ...props }, ref) => (
    <Container ref={ref} {...props}>
      <List>
        <ListItem>
          {items &&
            items.map((item, index) => {
              const iconProps = !!item.icon
                ? {
                    icon: item.icon,
                    iconPlacement: "left",
                    iconSize: "small",
                  }
                : {};
              return (
                <Button
                  key={index.toString()}
                  {...(iconProps as ButtonProps)}
                  onClick={item.onClick}
                  fontWeight={400}
                  variant={item.variant || ButtonVariant.Neutral}
                  fullWidth
                >
                  {item.text}
                </Button>
              );
            })}
        </ListItem>
      </List>
    </Container>
  )
);

Popup.displayName = "Popup";

export default memo(Popup);
