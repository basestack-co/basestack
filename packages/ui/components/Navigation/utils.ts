// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
