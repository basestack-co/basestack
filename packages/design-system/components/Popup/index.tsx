import { forwardRef, memo } from "react";
import type { PositionProps } from "styled-system";
import { Button, type ButtonProps, ButtonVariant } from "../Button";
import { Container, List, ListItem } from "./styles";

export interface PopupItemsProps {
  /**
   * onClick list item callback
   */
  onClick: () => void;
  /**
   * List item text
   */
  text: string;
  /**
   * Optional List item icon
   */
  icon?: string;
  /**
   * Change list item button styles based on variant
   */
  variant?: ButtonVariant;
  /**
   * Change the visibility of the item button
   */
  isVisible?: boolean;
  /**
   * Change the disability of the item button
   */
  isDisabled?: boolean;
}

export interface PopupProps extends PositionProps {
  /**
   * Callback onClick the list
   */
  onClickList?: () => void;
  /**
   * List of actions
   */
  items: Array<PopupItemsProps>;
  /**
   * Box width
   */
  width?: number;
}

const Popup = forwardRef<HTMLDivElement, PopupProps>(
  ({ items, width, onClickList, ...props }, ref) => (
    <Container width={width} ref={ref} {...props}>
      <List onClick={onClickList}>
        {items &&
          items.map(({ isVisible = true, ...item }, index) => {
            const iconProps = item.icon
              ? {
                  icon: item.icon,
                  iconPlacement: "left",
                  iconSize: "small",
                }
              : {};
            return isVisible ? (
              <ListItem key={index.toString()}>
                <Button
                  {...(iconProps as ButtonProps)}
                  onClick={item.onClick}
                  fontWeight={400}
                  variant={item.variant || ButtonVariant.Neutral}
                  isDisabled={item.isDisabled}
                  fullWidth
                >
                  {item.text}
                </Button>
              </ListItem>
            ) : null;
          })}
      </List>
    </Container>
  ),
);

Popup.displayName = "Popup";

export default memo(Popup);
