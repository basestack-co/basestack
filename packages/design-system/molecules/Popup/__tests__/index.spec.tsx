import React from "react";
import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Popup from "..";

describe("Popup Molecule tests", () => {
  afterEach(cleanup);

  test("should render Popup correctly", () => {
    const { asFragment } = renderWithTheme(
      <Popup
        position="absolute"
        right={0}
        top={0}
        items={[
          { text: "Edit", onClick: () => console.log("") },
          { text: "History", onClick: () => console.log("") },
          { text: "Delete", onClick: () => console.log("") },
        ]}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
