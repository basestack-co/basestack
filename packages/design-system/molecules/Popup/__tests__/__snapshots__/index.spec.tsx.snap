// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Molecule tests should render Popup correctly 1`] = `
<DocumentFragment>
  .c3 {
  border: none;
  height: 2.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 0.75rem;
  font-size: 0.875rem;
  font-weight: 500;
  cursor: pointer;
  border-radius: 4px;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: transparent;
  color: #000000;
  font-weight: 400;
  width: 100%;
}

.c3 .material-symbols-sharp {
  color: inherit;
}

.c3:hover {
  background-color: #EEEEEE;
}

.c3:active {
  background-color: #E2E2E2;
}

.c0 {
  position: absolute;
  right: 0;
  top: 0;
  background-color: #FFFFFF;
  box-shadow: rgba(0,0,0,0.25) 0px 25px 50px -12px;
  padding: 0.25rem;
  border-radius: 4px;
  width: 9.375rem;
  z-index: 101;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        <button
          class="c3"
          font-weight="400"
        >
          Edit
        </button>
        <button
          class="c3"
          font-weight="400"
        >
          History
        </button>
        <button
          class="c3"
          font-weight="400"
        >
          Delete
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
